spring-integration-core/src/main/java/org/springframework/integration/config/annotation/AbstractMethodAnnotationPostProcessor.java

```diff
@@ -106,67 +125,298 @@
 		this.conversionService = this.beanFactory.getConversionService() != null
 				? this.beanFactory.getConversionService()
 				: DefaultConversionService.getSharedInstance();
 		this.channelResolver = ChannelResolverUtils.getChannelResolver(beanFactory);
-		this.annotationType =
-				(Class<T>) GenericTypeResolver.resolveTypeArgument(this.getClass(),
-						MethodAnnotationPostProcessor.class);
+	}
+
+	protected ConfigurableListableBeanFactory getBeanFactory() {
+		return this.beanFactory;
+	}
+
+	protected BeanDefinitionRegistry getDefinitionRegistry() {
+		return this.definitionRegistry;
+	}
+
+	protected ConversionService getConversionService() {
+		return this.conversionService;
+	}
+
+	protected DestinationResolver<MessageChannel> getChannelResolver() {
+		return this.channelResolver;
 	}
 
 	@Override
-	public Object postProcess(Object bean, String beanName, Method method, List<Annotation> annotations) {
-		Object sourceHandler = null;
-		if (beanAnnotationAware() && AnnotatedElementUtils.isAnnotated(method, Bean.class.getName())) {
-			if (!this.beanFactory.containsBeanDefinition(resolveTargetBeanName(method))) {
-				this.logger.debug("Skipping endpoint creation; perhaps due to some '@Conditional' annotation.");
-				return null;
+	public void processBeanDefinition(String beanName, AnnotatedBeanDefinition beanDefinition,
+			List<Annotation> annotations) {
+
+		ResolvableType handlerBeanType = getHandlerBeanClass(beanDefinition);
+
+		String handlerBeanName = beanName;
+
+		BeanDefinition handlerBeanDefinition =
+				resolveHandlerBeanDefinition(beanName, beanDefinition, handlerBeanType, annotations);
+		MergedAnnotations mergedAnnotations = beanDefinition.getFactoryMethodMetadata().getAnnotations();
+
+		if (handlerBeanDefinition != null) {
+			if (handlerBeanDefinition != beanDefinition) {
+				Class<?> handlerBeanClass =
+						org.springframework.util.ClassUtils.resolveClassName(handlerBeanDefinition.getBeanClassName(),
+								this.beanFactory.getBeanClassLoader());
+
+				if (isClassIn(handlerBeanClass, Orderable.class, AbstractSimpleMessageHandlerFactoryBean.class)) {
+					mergedAnnotations.get(Order.class)
+							.getValue(AnnotationUtils.VALUE, String.class).
+							ifPresent((order) -> handlerBeanDefinition.getPropertyValues().add("order", order));
+				}
+
+				if (isClassIn(handlerBeanClass, AbstractMessageProducingHandler.class, AbstractMessageRouter.class,
+						AbstractSimpleMessageHandlerFactoryBean.class)) {
+
+					new BeanDefinitionPropertiesMapper(handlerBeanDefinition, annotations)
+							.setPropertyValue(SEND_TIMEOUT_ATTRIBUTE)
+							.setPropertyReference("outputChannel")
+							.setPropertyReference("defaultOutputChannel");
+				}
+
+				if (isClassIn(handlerBeanClass, AbstractMessageProducingHandler.class,
+						AbstractSimpleMessageHandlerFactoryBean.class)
+						&& !RouterFactoryBean.class.isAssignableFrom(handlerBeanClass)) {
+
+					applyAdviceChainIfAny(handlerBeanDefinition, annotations);
```

Comment:
Why are we doing this here - this will add all the advices to the handler - the CEFB (which also gets the full advice chain) selectively applies advices to the handler.

Reply:
See the comment exactly in that `ConsumerEndpointFactoryBean.adviceChain()`:
```
 *  ARPMHs advise the handleRequestMessage method internally and already have the advice chain injected.
*  So we only advise handlers that are not reply-producing.
```

so, we don't set an `adviceChain` to the `AbstractReplyProducingMessageHandler`, neither from XML parser, nor Java DSL - `ConsumerEndpointSpec.doGet()`.

We may revise it though, but I guess there was a reason why you implemented that way in the XML parser day first.
Different issue either way.